src/Chainweb/Pact/PactService.hs

```diff
@@ -477,97 +478,134 @@ finalizeCheckpointer finalize = do
 _liftCPErr :: Either String a -> PactServiceM cas a
 _liftCPErr = either internalError' return
 
-type Balances = HM.HashMap Text Decimal
 
 data Uniqueness = Duplicate | Unique
 
+-- | Performs a dry run of PactExecution's `buyGas` function for transactions being validated.
+--
+attemptBuyGas
+    :: PayloadCas cas
+    => Checkpointer
+    -> Miner
+    -> Maybe BlockHash
+    -> Vector (ChainwebTransaction, Uniqueness)
+    -> PactServiceM cas (Vector (ChainwebTransaction, Uniqueness, Bool))
+attemptBuyGas cp miner nonGenesisParentHash txs = withDiscardedBatch $ do
+  mbLatestBlock <- liftIO $ _cpGetLatestBlock cp
+  (bh, bhash) <- case mbLatestBlock of
+                     Nothing -> throwM NoBlockValidatedYet
+                     (Just !p) -> return p
+  let target = Just (succ bh, bhash)
+  withCheckpointer target "validateChainwebTxs/attemptBuyGas" $ \(PactDbEnv' dbEnv) -> do
+      (T2 _ mcache) <- runCoinbase nonGenesisParentHash dbEnv miner
```

Comment:
Is this to catch miner problems?

Reply:
`buyGas` takes a ModuleCache as an input. And after tracing back to see where it got it from, I came across these:
- https://github.com/kadena-io/chainweb-node/blob/master/src/Chainweb/Pact/TransactionExec.hs#L164
-> https://github.com/kadena-io/chainweb-node/blob/masg/src/Chainweb/Pact/PactService.hs#L938
-> https://github.com/kadena-io/chainweb-node/blob/masg/src/Chainweb/Pact/PactService.hs#L878
(it seems that the `ModuleCache` that eventually gets run with `runGas` get created in the latter via: `T2 coinOut mc <- runCoinbase nonGenesisParentHash pactdbenv miner`. So I tried to replicate this.)

Reply:
call to `runCoinbase` omitted in current commit.